server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBody.java

```diff
@@ -0,0 +1,107 @@
+ */
+public interface ChunkedRestResponseBody {
+
+    /**
+     * @return true once this response has been written fully.
+     */
+    boolean isDone();
+
+    /**
+     * Serializes approximately as many bytes of the response as request by {@code sizeHint} to a {@link ReleasableBytesReference} that
+     * is created from buffers backed by the given {@code recycler}.
+     *
+     * @param sizeHint how many bytes to approximately serialize for the given chunk
+     * @param recycler recycler used to acquire buffers
+     * @return serialized chunk
+     * @throws IOException on serialization failure
+     */
+    ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) throws IOException;
+
+    /**
+     * Create a chunked response body to be written to a specific {@link RestChannel} from a {@link ChunkedToXContent}.
+     *
+     * @param chunkedToXContent chunked x-content instance to serialize
+     * @param params parameters to use for serialization
+     * @param channel channel the response will be written to
+     * @return chunked rest response body
+     */
+    static ChunkedRestResponseBody fromXContent(ChunkedToXContent chunkedToXContent, ToXContent.Params params, RestChannel channel)
+        throws IOException {
+
+        return new ChunkedRestResponseBody() {
+
+            private final OutputStream out = new OutputStream() {
+                @Override
+                public void write(int b) throws IOException {
+                    target.write(b);
+                }
+
+                @Override
+                public void write(byte[] b, int off, int len) throws IOException {
+                    target.write(b, off, len);
+                }
+            };
+
+            private final XContentBuilder builder = channel.newBuilder(
+                channel.request().getXContentType(),
+                null,
+                true,
+                Streams.noCloseStream(out)
+            );
+
+            private final Iterator<ToXContent> serialization = chunkedToXContent.toXContentChunked();
+
+            private BytesStream target;
+
+            @Override
+            public boolean isDone() {
+                return serialization.hasNext() == false;
+            }
+
+            @Override
+            public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) throws IOException {
+                final RecyclerBytesStreamOutput chunkStream = new RecyclerBytesStreamOutput(recycler);
+                assert this.target == null;
+                this.target = chunkStream;
+                while (serialization.hasNext()) {
+                    serialization.next().toXContent(builder, params);
+                    if (chunkStream.size() >= sizeHint) {
+                        break;
+                    }
+                }
+                if (serialization.hasNext() == false) {
```

Comment:
You don't have to flush the builder manually, closing should flush everything from it. We don't really limit the size of the underlying stream as well so even if we overshoot here it's not so bad except for a bit of wasted memory?
Maybe I'm getting the question wrong?

Reply:
Sorry for missing this comment in my previous review. I were not after the case where we close (which obviously flushes too), rather the case where we do not close it. That would ensure a "whole" object is returned per chunk rather than a chunk splitting across json. Might not be a big deal at all and I'd be ok to omit this, just feels nicer.

Reply:
Ah understood. I'm not sure we actually want this. In the end, the current state of affairs where the chunks are somewhat aligned with the JSON objects is more of a side-effect of not wanting to make this too complicated in step 1. It seems like we'd rather want to size the chunks as evenly as possible to avoid overhead from too many chunks on the one hand and get smooth memory usage on the other hand.

I was to some degree planning to do a follow-up here that adds a buffering layer in the pipeline to give us nice and evenly sized chunks if I ever found the time :)